File: apps/docs/src/prework/examples/structural.ts
type ZoneId = string;
type SegmentId = string;

const segmentId: SegmentId = "SEG-07";
// 別名を付けても、どちらも string。代入は通る。
export const mistakenZoneId: ZoneId = segmentId;

type Zone = Readonly<{ kind: "ZoneId"; value: string }>;
type Segment = Readonly<{ kind: "SegmentId"; value: string }>;

export function readZone(zone: Zone): string {
  return zone.value;
}

export function typeExamples(segment: Segment) {
  // @ts-expect-error: kind が異なるため、Zone として渡せない。
  readZone(segment);
}

File: apps/docs/src/prework/examples/rust/structural.rs
pub type ZoneId = String;
pub type SegmentId = String;

pub fn mistaken_zone_id(segment: SegmentId) -> ZoneId {
    segment // 別名だけでは区別できない。
}

// 同じフィールドを持っていても別のstructは別の型。
pub struct Zone(pub String);
pub struct Segment(pub String);

pub fn read_zone(zone: &Zone) -> &str {
    &zone.0
}

// read_zone(&Segment("SEG-07".into())) はコンパイルエラー。
